Added new sklearn models and hyper-paremeter tuning for classification workflow.

dickeygh opened the pull request:
Get_Task_Type.py:
1. Added definition of tune variable to be used in main.py.
2. Adjusted to work with new classification models.

Classifiers.py:
1. Redid parts of this file to allow for hyper-parameter tuning.
2. Added extra models

features.py:
1. Added an if statement to allow classification workflow to run.

grid.py:
1. Added new tuning for linearSVC

name.py:
1. Added linearSVC to algorithm_list

regressors.py:
1. Added if statements at lines 161-164 and lines 212-216 to allow for hyper-tuning in the classification workflow.

Issues 82 and 84 are related to this PR. Please check them out and let me know of any potential solutions. However, these issues do not cause problems with this PR, as Get_Task_Type.py allows the workflow to run without these errors being raised.

pep8speaks commented:
Hello @dickeygh! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`core/Get_Task_Type.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py):

> [Line 11:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (163 > 79 characters)
> [Line 15:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L15): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 19:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L19): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 79 characters)
> [Line 37:5](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L37): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 37:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L37): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 79 characters)
> [Line 38:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L38): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 43:36](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/Get_Task_Type.py#L43): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`core/classifiers.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/classifiers.py):

> [Line 3:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/classifiers.py#L3): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 5:1](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/classifiers.py#L5): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

* In the file [`core/features.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/features.py):

> [Line 35:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/features.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)



* In the file [`core/name.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/name.py):

> [Line 69:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/name.py#L69): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)

* In the file [`core/regressors.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/regressors.py):

> [Line 162:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/regressors.py#L162): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 222:80](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/core/regressors.py#L222): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)

* In the file [`main.py`](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/main.py):

> [Line 22:4](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/main.py#L22): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 27:4](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/main.py#L27): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 44:10](https://github.com/Dlux804/McQuade-Chem-ML/blob/a4dbf84899b01f1728b32b84ff6a229d21ca2e30/main.py#L44): [E131](https://duckduckgo.com/?q=pep8%20E131) continuation line unaligned for hanging indent

qle2 commented on core/Get_Task_Type.py:
```diff
@@ -8,15 +8,15 @@
-    if data in ['sider.csv', 'clintox.csv'] and alg == 'svm':  # SVC is not compatible with multi-label classification.
+    if data in ['sider.csv', 'clintox.csv'] and alg in ['svm', 'linearSVC', 'ada', 'gdb']:  # SVC, ada, and gdb are not compatible with multi-label classification.
         checker = 0
 
         # These classification data sets are not compatible with these regression models
-    elif data in ['BBBP.csv', 'sider.csv', 'clintox.csv', 'bace.csv'] and alg in ['ada', 'gdb', 'nn']:
+    elif data in ['BBBP.csv', 'sider.csv', 'clintox.csv', 'bace.csv'] and alg in ['nn']:
         checker = 0
 
         # These regression data sets are not compatible with these classification models.
-    elif data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv', 'jak2_pic50.csv'] and alg in []:
+    elif data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv', 'jak2_pic50.csv'] and alg in ['linearSVC']:
         checker = 0
 
     else: # Checker = 1 when there are no compatibility
```
My bad for suggesting `linearSVC` for multi-label, we should remove it. Sklearn has a list for all the [algorithms that support multi-label classification](https://scikit-learn.org/stable/modules/multiclass.html). Since we already have `MLP` and `KNN` in our pipeline, it should be an easy add. They may require different hyperparameters.

qle2 commented on core/name.py:
```diff
@@ -66,7 +66,7 @@ def represent_algorithm():
     :return:
     """
-    algorithm_list = ['ada', 'rf', 'gdb', 'mlp', 'knn', 'nn', 'svm']  # Out current supported algorithm
+    algorithm_list = ['ada', 'rf', 'gdb', 'mlp', 'knn', 'nn', 'svm', 'linearSVC']  # Out current supported algorithm
```
We can remove `linearSVC` since it does not support multi-label classification according to sklearn.

Dlux804 replied:
Does it provide any benefit?

qle2 replied:
No, it doesn't. I thought `linearSVC` can be used for multi-label not it's not supported.

qle2 commented on core/regressors.py:
```diff
@@ -167,7 +172,7 @@ def hyperTune(self, epochs=50, n_jobs=6):
         n_iter=self.opt_iter,  # number of combos tried
         random_state=42,  # random seed
         verbose=3,  # output print level
-        scoring='neg_mean_squared_error',  # scoring function to use (RMSE)  #TODO needs update for Classification
+        scoring=scoring,  # scoring function to use (RMSE)
         n_jobs=n_jobs,  # number of parallel jobs (max = folds)
         cv=self.cv_folds  # number of cross-val folds to use
```
Good add

qle2 commented on main.py:
```diff
@@ -19,28 +19,34 @@ def main():
 
     # list of available classification learning algorithms for reference/testing
-    # learner = ['svm', 'knn', 'rf']
+   # learner = ['linearSVC', 'svm', 'knn', 'rf', 'ada', 'gdb']
```
Remove `linearSVC`. Refer to the comments above.

qle2 reviewed (commented):
Looks good. I would remove `linearSVC` since it does not support multi-label classification. If we want more [algorithms for multi-label](https://scikit-learn.org/stable/modules/multiclass.html), I suggest adding `KNN` and `MLP`.

Dlux804 commented on core/Get_Task_Type.py:
```diff
@@ -8,15 +8,15 @@
 
         # These regression data sets are not compatible with these classification models.
-    elif data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv', 'jak2_pic50.csv'] and alg in []:
+    elif data in ['ESOL.csv', 'Lipophilicity-ID.csv', 'water-energy.csv', 'logP14k.csv', 'jak2_pic50.csv'] and alg in ['linearSVC']:
```
I don't understand these `elif` statements for regression/classification datasets and algorithms.  You are saying that `nn` is only for regression and `linearSVC` is only for classification, right?

dickeygh replied:
Yes, that is correct.

Dlux804 commented on core/regressors.py:
```diff
@@ -159,6 +159,11 @@ def hyperTune(self, epochs=50, n_jobs=6):
         self.fit_params = None  # sklearn models don't need fit params
 
+    if self.task_type in ['single_label_classification', 'multi_label_classification']:
+        scoring = None
```
`scoring = None` sets to default, yes?  What is default?

qle2 replied:
Setting scoring to None will set the scoring function to whatever is the scoring method for each algorithm. For example, the default scoring method for RandomForestRegressor is R2

dickeygh replied:
I believe the default for most classifiers is accuracy score or AUC.

Dlux804 replied:
We should make sure that we is the same for each classifier -- either accuracy or AUC and use it for all classifiers.

dickeygh replied:
I did some research, and it seems that every classifier we are using uses accuracy score as its default scoring method.

Dlux804 commented on main.py:
```diff
@@ -73,7 +79,7 @@ def main():
                     target = get_classification_targets(data)
```
Why not just set this up in the dictionary?  Why set to `None` and then define later with an `if`?

dickeygh replied:
Setting this up in the dictionary caused errors later on in the classification workflow that I could not solve. I believe this had to do with the target list being appended later in the workflow.

Dlux804 commented on main.py:
```diff
@@ -88,7 +94,7 @@ def main():
 
                         model = MlModel(algorithm=alg, dataset=data, target=target, feat_meth=method,
-                                        tune=True, cv=2, opt_iter=2)
+                                        tune=tune, cv=2, opt_iter=2)
```
I'm not a fan of this `tune=tune`.  I get the reasoning -- some of the algorithms are not supported.  But this removes the option of having tune = False for algorithms which are supported.

dickeygh replied:
I see your point. Any suggestions for a work-around?

Dlux804 replied:
The simplest that comes to mind quickly is to apply the filter after setting the `tune` variable.  Right now we filter and then set `tune`, removing the potential for setting it to false for algorithms it works for.  However, if we apply the filter after we set a preference, we can overwrite the `tune` for the ones it does not work for.  For example, a user could set `tune=True` and then if they are using an unsupported algorithm, `tune` will be overwritten to `False`.

Dlux804 commented on core/features.py:
```diff
@@ -31,7 +31,9 @@ def featurize(self, not_silent=True, retrieve_from_mysql=False):
     feat_meth = self.feat_meth
     df = self.data
-    df['smiles'] = canonical_smiles(list(df['smiles']))  # Turn SMILES into CANONICAL SMILES
+    if self.task_type == 'regression':
+        df['smiles'] = canonical_smiles(list(df['smiles']))  # Turn SMILES into CANONICAL SMILES
```
Why would this be a regression only activity?  This is a must for all molecular datasets.

dickeygh replied:
I added this if statement because it wasn't working for the classification workflow. I will work on getting this to work for all data sets.

Dlux804 reviewed (changes requested):
The code runs, which is good.  It is rough around the edges as it goes through a lot of `if else` to avoid errors.  Most of these issues can be smoothed out as we go on if the models are creating good results.  To that end, there are some things that must be addressed. 

- Canonicalization in `features.py` is not just for regression.  
- Specifying whether or not to tune a model in `main.py` needs to remain flexible for easy prototyping
- Clean up/enhance output of model predictions and metrics. Some graphs are missing axis titles.  Predictions need to be capture per molecule, etc.

dickeygh commented:
Today, I pushed several changes to this PR:

 Classifiers.py:
1. Removed linearSVC

Get_Task_Type.py:
1. Removed linearSVC
2. Removed section for deciding if tuning should be performed.

grid.py:
1. Removed linearSVC

models.py:
1. Added new if statement for making tune = false for the classification models that can't tune properly.

name.py:
1. Removed linearSVC

main.py:
1. Removed linearSVC
2. Changed how tune is decided so that it can be set to false or true.



Features.py:
1. Moved lines 73-88 to lines 36-50. This is so that canonicalization will work properly for the classification workflow.

qle2 commented on core/regressors.py:
```diff
@@ -159,6 +159,11 @@ def hyperTune(self, epochs=50, n_jobs=6):
         self.fit_params = None  # sklearn models don't need fit params
 
+    if self.task_type in ['single_label_classification', 'multi_label_classification']:
+        scoring = None
```
Find a metric that you believe is a good reflection of the classification accuracy and use it for all of your classifiers

qle2 reviewed (dismissed):
Good fix. Find a metric for all classfication

qle2 commented on core/models.py:
```diff
@@ -43,6 +43,8 @@ def __init__(self, algorithm, dataset, target, feat_meth, tune=False, opt_iter=1
 
         self.algorithm = algorithm
         self.dataset = dataset
+        if self.algorithm in ['svm', 'ada'] and self.dataset in ['BBBP.csv', 'sider.csv', 'clintox.csv', 'bace.csv']:
+            tune = False
```
One final question here: Why is tune set to `false` for SVM and ADA for all four dataset? I thought it was easy with `BBBP.csv` and `base.csv` since they are single label?

dickeygh replied:
SVM classifier and ADA classifier don't work properly with hyper tuning regardless of multi/single label. Issue #84 contains these error messages. I have not been able to find a fix for this.

qle2 reviewed (commented):
Why is tune set to `False` with SVM and ADA for all classification dataset?

dickeygh commented:
Today, I made the following changes to this PR:

1. I merged it with the newly updated development branch



 Analysis.py:
1. Used self.predictions_analysis instead so that the classification workflow would save/analyze correctly for both multi-label and single-label.

Train.py:
1. Added a split between multi and single label, and added declaration of self.predictions_analysis to be used for single_label analysis.

Dlux804 commented on core/train.py:
```diff
@@ -235,14 +235,36 @@ def train_cls(self, n=5):
-    print('Average f1_score = %.3f' % stats['f1_score_avg'], '+- %.3f' % stats['f1_score_std'])
-    print()
+    if self.task_type == 'single_label_classification':
+        self.predictions = cls_multi[target_name]
+        self.predictions_analysis = predictions
+        self.predictions_stats = stats
```
@dickeygh This is confusing that you store `cls_multi[target_name]` to `self.predictions` and then store `predictions` to `self.prediction_analysis`.  What is the difference between `cls_multi[target_name` , `predictions`, and `stats`?

Dlux804 reviewed (changes requested):
When I run it locally, `main.py main()` it fails during the neo4j import.  

```
Traceback (most recent call last):
  File "C:/Users/Adam/OneDrive/Research/Graduate/Projects/NSF/OKN/phase1/Work/scripting/McQuade-Chem-ML/main.py", line 225, in <module>
    main()
  File "C:/Users/Adam/OneDrive/Research/Graduate/Projects/NSF/OKN/phase1/Work/scripting/McQuade-Chem-ML/main.py", line 114, in main
    model.to_neo4j(port="bolt://localhost:7687", username="neo4j", password="password")
  File "C:\Users\Adam\OneDrive\Research\Graduate\Projects\NSF\OKN\phase1\Work\scripting\McQuade-Chem-ML\core\models.py", line 140, in to_neo4j
    ModelToNeo4j(model=self, port=port, molecules_per_batch=5000, username=username, password=password)
  File "C:\Users\Adam\OneDrive\Research\Graduate\Projects\NSF\OKN\phase1\Work\scripting\McQuade-Chem-ML\core\neo4j\models_to_neo4j.py", line 63, in __init__
    self.parsed_models = self.initialize_model_object()
  File "C:\Users\Adam\OneDrive\Research\Graduate\Projects\NSF\OKN\phase1\Work\scripting\McQuade-Chem-ML\core\neo4j\models_to_neo4j.py", line 188, in initialize_model_object
    all_raw_dfs['scaled_predictions'])
KeyError: 'scaled_predictions'

Process finished with exit code 1
```
It's looking for the scaled predictions dataframe that apparently doesn't exists. Line that throws the error is:
 ```
 model_data = self.__combine_model_data_with_predictions__(all_raw_dfs['data'], all_raw_dfs['predictions'],
                                                                  all_raw_dfs['scaled_predictions'])
```
 @andreshyer did you address this in your last PR?

Dlux804 reviewed (approved):
I have run this locally and it works so far.  Approved.
